src/macros.rs

```diff
@@ -87,3 +87,17 @@ macro_rules! log_func {
         &name[..name.len() - 3]
     }};
 }
+
+#[macro_export]
+macro_rules! pickle_extract {
+    ($self: ident, $pydict:expr, $attr_name:ident) => {
+        $self.$attr_name = $pydict
+            .get_item(stringify!($attr_name))
+            .ok_or_else(|| {
+                PyValueError::new_err(format!(
+                    "bad pickle contents for {}: {}",
+                    stringify!($attr_name), $pydict
+                ))
+            })?.extract()?;
+    };
+}
```

Comment:
We could do this with a function, which I'd prefer over a macro:

```rust
fn pickle_extract<'a, D>(dict: &'a PyDict, key: &str) -> PyResult<D>
where
    D: FromPyObject<'a>,
{
    dict.get_item(key)
        .ok_or_else(|| PyValueError::new_err(format!("bad pickle contents for {}: {}", key, dict)))?
        .extract()
}
```

And used like this:
```rust
        self.service_name = pickle_extract(dict, "service_name")?;

```